pandas/_libs/src/vendored/numpy/datetime/np_datetime.c

```diff
@@ -338,6 +338,31 @@ PyObject *extract_utc_offset(PyObject *obj) {
   return tmp;
 }
 
+static int scaleAndOffset(int64_t timestamp, int64_t scale, int64_t offset,
```

Comment:
I don't think the scaling functions should change at all. Where is the overflow in your example actually occurring?

Reply:
I'm using the following code to trigger the overflow.

```python
import pandas as pd
pd.DatetimeIndex([pd.Timestamp.min.tz_localize("utc")])

Traceback (most recent call last):
  File "conversion.pyx", line 501, in pandas._libs.tslibs.conversion.convert_datetime_to_tsobject
OverflowError: Overflow occurred in npy_datetimestruct_to_datetime

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/rschmidtke/workspace/other/pandas/overflow.py", line 10, in <module>
    main()
  File "/home/rschmidtke/workspace/other/pandas/overflow.py", line 7, in main
    print(pd.DatetimeIndex([pd.Timestamp.min.tz_localize("utc")]))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rschmidtke/miniforge3/envs/pandas/lib/python3.11/site-packages/pandas/core/indexes/datetimes.py", line 370, in __new__
    dtarr = DatetimeArray._from_sequence_not_strict(
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rschmidtke/miniforge3/envs/pandas/lib/python3.11/site-packages/pandas/core/arrays/datetimes.py", line 380, in _from_sequence_not_strict
    subarr, tz = _sequence_to_dt64(
                 ^^^^^^^^^^^^^^^^^^
  File "/home/rschmidtke/miniforge3/envs/pandas/lib/python3.11/site-packages/pandas/core/arrays/datetimes.py", line 2272, in _sequence_to_dt64
    converted, inferred_tz = objects_to_datetime64(
                             ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/rschmidtke/miniforge3/envs/pandas/lib/python3.11/site-packages/pandas/core/arrays/datetimes.py", line 2422, in objects_to_datetime64
    result, tz_parsed = tslib.array_to_datetime(
                        ^^^^^^^^^^^^^^^^^^^^^^^^
  File "tslib.pyx", line 414, in pandas._libs.tslib.array_to_datetime
  File "tslib.pyx", line 596, in pandas._libs.tslib.array_to_datetime
  File "tslib.pyx", line 505, in pandas._libs.tslib.array_to_datetime
  File "conversion.pyx", line 799, in pandas._libs.tslibs.conversion.parse_pydatetime
  File "conversion.pyx", line 504, in pandas._libs.tslibs.conversion.convert_datetime_to_tsobject
pandas._libs.tslibs.np_datetime.OutOfBoundsDatetime: Out of bounds nanosecond timestamp, at position 0
```

I have traced the individual scale operations for this below and added comments.

```c
npy_datetime npy_datetimestruct_to_datetime(NPY_DATETIMEUNIT base,
                                            const npy_datetimestruct *dts) {
// base: 10 (NPY_FR_ns)
// dts->year: 1677
// dts->month: 9
// dts->day: 21
// dts->hour: 0
// dts->min: 12
// dts->sec: 43
// dts->us: 145224
// dts->ps: 193000
// dts->as: 0

/* ... snip ... */

const int64_t days = get_datetimestruct_days(dts);
// days: -106752
// this is 1677-09-21 00:00:00, which is below the minimum timestamp and eventually causes the overflow during scaling

/* ... snip ... */

int64_t hours;
PD_CHECK_OVERFLOW(scaleDaysToHours(days, &hours));
PD_CHECK_OVERFLOW(checked_int64_add(hours, dts->hour, &hours));
// hours: -2562048

/* ... snip ... */

int64_t minutes;
PD_CHECK_OVERFLOW(scaleHoursToMinutes(hours, &minutes));
PD_CHECK_OVERFLOW(checked_int64_add(minutes, dts->min, &minutes));
// minutes: -153722868

/* ... snip ... */

int64_t seconds;
PD_CHECK_OVERFLOW(scaleMinutesToSeconds(minutes, &seconds));
PD_CHECK_OVERFLOW(checked_int64_add(seconds, dts->sec, &seconds));
// seconds: -9223372037

/* ... snip ... */

int64_t microseconds;
PD_CHECK_OVERFLOW(scaleSecondsToMicroseconds(seconds, &microseconds));
PD_CHECK_OVERFLOW(checked_int64_add(microseconds, dts->us, &microseconds));
// microseconds: -9223372036854776

/* ... snip ... */

if (base == NPY_FR_ns) {
  int64_t nanoseconds;

  // the following overflows because -9223372036854776 * 1000 = -9223372036854776000,
  // but NPY_MIN_INT64 is -9223372036854775808
  PD_CHECK_OVERFLOW(
    scaleMicrosecondsToNanoseconds(microseconds, &nanoseconds));

  // the following would put the nanoseconds back into the valid range
  // -9223372036854776000 + 193 = -9223372036854775807
  PD_CHECK_OVERFLOW(
    checked_int64_add(nanoseconds, dts->ps / 1000, &nanoseconds));

  return nanoseconds;
}
}
```

For positive timestamps (= after 1/1/1970), the scaling + addition approach always works, because they both produce increasing values. However, for negative timestamps, the scaling operation produces a decreasing value, while the addition increases it again. Hence, even if the final timestamp is valid, the scaling may overflow. Hence I think the two operations (scaling + addition) have to be done in tandem, because the scaling overflow could be remedied by the addition.

Alternatively, one could (for negative timestamps), add one day to `days` and subtract it at the end for every returning branch:

```c
int64_t days = get_datetimestruct_days(dts);
int negative = 0;
if (days < 0) {
  ++days;
  negative = 1;
}

/* ... snip ... */

if (base == NPY_FR_ns) {
  int64_t nanoseconds;
  PD_CHECK_OVERFLOW(
    scaleMicrosecondsToNanoseconds(microseconds, &nanoseconds));
  PD_CHECK_OVERFLOW(
    checked_int64_add(nanoseconds, dts->ps / 1000, &nanoseconds));

  if (negative) {
    // subtract one day in nanoseconds
    PD_CHECK_OVERFLOW(
      checked_int64_sub(nanoseconds, 86400000000000, &nanoseconds));
  }

  return nanoseconds;
}
```

Reply:
I prefer the second idea you present to just check for values at the negative extreme before scaling to nanoseconds. I think the code you prototyped is still _slightly_ off because it will overflow before you check for `negative`, but I think it is headed in the right direction

Reply:
Sounds good, I'll try and come up with a maintainable solution to that over the course of this week.